src/libre/ast_analysis.c

```diff
@@ -173,6 +173,10 @@ always_consumes_input(const struct ast_expr *n, int thud)
 	case AST_EXPR_NAMED:
 		return 1;
 
+	case AST_EXPR_SUBTRACT:
+		/* FIXME: need to perform the subtraction to know */
+		return 1;
+
```

Comment:
This worries me a bit. Is there ever a correct way to find this here?

I don't see how to avoid subtraction in the AST given that at least one dialect has an operator for it.

Reply:
The main edge case here is if the subtraction gives the empty set.

Reply:
Can we have a test which hits this case? I think that'd be either in the SQL dialect using `[x^x]` (to produce an empty set) or in the native dialect using `[^\x00-\xff]` to achieve the same.

Reply:
Indeed these both fail on master:
```
; bmake -r CC=gcc DEBUG=1 && ./build/bin/re -r sql -pl dot '[x^x]'
re: src/libre/re.c:169: re_comp: Assertion `err == NULL || err->e != RE_ESUCCESS' failed.
zsh: abort      ./build/bin/re -r sql -pl dot '[x^x]'
; bmake -r CC=gcc DEBUG=1 && ./build/bin/re -r native -pl dot '[^\x00-\xff]'
re: src/libre/re.c:169: re_comp: Assertion `err == NULL || err->e != RE_ESUCCESS' failed.
zsh: abort      ./build/bin/re -r native -pl dot '[^\x00-\xff]'
; 
```

Reply:
Okay, I addressed these assertions in #187, so we should be able to use `[^\x00-\xff]` or equivalent in the contexts which'd reach this `/* FIXME */`.